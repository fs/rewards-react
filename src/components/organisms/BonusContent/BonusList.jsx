import React from 'react';
import styled from 'styled-components';
import Bonus from '../../molecules/Bonus';

const List = styled.div``;

const Loader = styled.div`
  margin: 20px;
  text-align: center;
`;

const BonusList = (props) => {
  const { bonusList, isLoading } = props;
  return (
    <div>
      { isLoading
        ? (
          <Loader data-testid="test-loader">
            Content is loading
          </Loader>
        )
        : (
          <List data-testid="test-bonus-list">
            {bonusList.map(bonus => <Bonus bonus={bonus} key={bonus.id} />)}
          </List>
        )
      }
    </div>
  );
};

export default BonusList;
