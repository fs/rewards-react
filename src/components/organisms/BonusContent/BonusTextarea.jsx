import React from 'react';
import ReactTextareaAutocomplete from '@webscopeio/react-textarea-autocomplete';
import '../../../styles/autocomplete.css';
import BonusPossibilitiesService, {
  POINTS, TAGS, USERS,
} from '../../../services/BonusPossibilitiesService';

const BonusTextarea = (props) => {
  const { onChange } = props;
  const pointItem = ({ entity: { id, value } }) => <div data-testid="test-point-item" key={id}>{`+â™¥${value}`}</div>;
  const userItem = ({ entity: { id, username } }) => <div data-testid="test-user-item" key={id}>{`@${username}`}</div>;
  const tagItem = ({ entity: { id, label } }) => <div data-testid="test-tag-item" key={id}>{`#${label}`}</div>;

  const points = BonusPossibilitiesService.getPossibilities(POINTS);
  const users = BonusPossibilitiesService.getPossibilities(USERS);
  const tags = BonusPossibilitiesService.getPossibilities(TAGS);

  return (
    <div data-testid="test-textarea-wrapper">
      <ReactTextareaAutocomplete
        className="autocomplete-textarea"
        onChange={onChange}
        placeholder="+100 @person add description for #create_awesomness"
        data-testid="test-textarea"
        loadingComponent={() => <span>Loading</span>}
        name="bonus-text"
        trigger={{
          '+': {
            dataProvider: token => points.filter(point => point.id.includes(token)),
            component: pointItem,
            output: (item, trigger) => `${trigger}${(item.value).toString()}`,
          },
          '@': {
            dataProvider: token => users.filter(user => user.username.includes(token)),
            component: userItem,
            output: (item, trigger) => `${trigger}${item.username}`,
          },
          '#': {
            dataProvider: token => tags.filter(tag => tag.label.includes(token)),
            component: tagItem,
            output: (item, trigger) => `${trigger}${item.label}`,
          },
        }}
        minChar="0"
      />
    </div>
  );
};

export default BonusTextarea;
