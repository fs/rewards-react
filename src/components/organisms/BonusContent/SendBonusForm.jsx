import React, { useState } from 'react';
import styled from 'styled-components';
import Button from '../../atoms/Button';
import authService from '../../../services/AuthService';
import bonusService from '../../../services/BonusService';

const Form = styled.form`
  width: 100%;
  background-color: #fff;
  border-radius: 5px;
  margin: 0 0 20px;
  padding: 20px;
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.1);
`;

const Textarea = styled.textarea`
  display: block;
  width: 100%;
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.75rem;
  color: #000;
  border: none;
  outline: none;
  padding: 0;
  height: 3.5rem;
  margin: 0 0 20px;
  resize: none;
  box-shadow: none;
`;

const SendBonusForm = (props) => {
  const [bonusText, setBonusText] = useState('');
  const [hasError, setHasError] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');

  const handleChange = (event) => {
    setBonusText(event.target.value);
    setHasError(false);
    setErrorMessage('');
  };

  const handleSubmit = async (event) => {
    event.preventDefault();
    const { onSuccess } = props;
    const token = authService.getToken();
    try {
      await bonusService.createBonus(token, bonusText);
      onSuccess();
    } catch (error) {
      const parsedErrorMessage = JSON.parse(error.response.request.response).errors[0].detail;
      setHasError(true);
      setErrorMessage(parsedErrorMessage);
    }
  };

  return (
    <Form onSubmit={handleSubmit} data-test="test-form">
      <Textarea
        name="bonustext"
        onChange={handleChange}
        placeholder="+100 @person add description for #create_awesomness"
        data-testid="test-textarea"
      />
      {hasError && <div className="error-message">{errorMessage}</div>}
      <Button text="Give" data-testid="test-button" />
    </Form>
  );
};

export default SendBonusForm;
