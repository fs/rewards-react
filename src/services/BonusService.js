import api from './ApiService';

const apiUrl = '/user/bonuses';

export default class BonusService {
  static async createBonus(token, text) {
    const config = {
      headers: { Authorization: `Bearer ${token}` },
    };

    const bodyParameters = {
      data: {
        type: 'bonus-texts',
        attributes: {
          text,
        },
      },
    };

    const response = await api.post(apiUrl, bodyParameters, config);
    return response;
  }

  static async updateBonusesList(token) {
    const config = {
      headers: { Authorization: `Bearer ${token}` },
    };

    const response = await api.get(apiUrl, config);
    return response;
  }
}
