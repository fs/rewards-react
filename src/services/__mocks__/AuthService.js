export default class AuthService {
  static async authenticate() {
    console.log('authenticate');
  }

  static async fetchToken() {
    return 'token';
  }

  static getToken() {
    return 'token';
  }
}
