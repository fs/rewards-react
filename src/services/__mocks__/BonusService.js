export default class BonusService {
  static async createBonus() {
    console.log('createBonus');
    return {};
  }

  static async fetchBonusesList() {
    console.log('fetchBonusesList');
    return {};
  }
}
