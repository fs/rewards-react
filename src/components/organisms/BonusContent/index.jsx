import React, { useState, useEffect, useContext } from 'react';
import styled from 'styled-components';
import autosize from 'autosize';
import SendBonusForm from '../SendBonusForm';
import BonusList from './BonusList';
import Context from '../../pages/BonusesPage/Context';
import bonusService from '../../../services/BonusService';
import authService from '../../../services/AuthService';
import bonusParser from '../../../utils/bonusParser';
import profileService from '../../../services/ProfileService';

const BonusContentWrapper = styled.div`
  position: relative;
  margin-bottom: 1.25rem;
  margin-right: 2rem;
  flex: 1;
  border-radius: 4px;
`;

const MyBonuses = styled.h2`
  position: absolute;
  top: -55px;
  line-height: 1.625rem;
  font-weight: 800;
  font-size: 1.375rem;
  color: #fff;
`;

const BonusContent = () => {
  const [bonusList, setBonusList] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const [hasError, setHasError] = useState(false);

  const { state, dispatch } = useContext(Context);

  const updateBonusesList = async () => {
    setHasError(false);
    setIsLoading(true);
    const token = authService.getToken();
    try {
      const bonusListObject = await bonusService.fetchBonusesList(token);
      const bonusListArray = bonusParser(bonusListObject.data);

      setBonusList(bonusListArray);

      // const userData = await profileService.fetchUser();
      // setCurrentUser(userData.data);
    } catch (error) {
      setHasError(true);
    }
    setIsLoading(false);

    autosize(document.querySelectorAll('textarea'));
  };

  let currentUser;

  const updateCurrentUser = async () => {
    try {
      const userData = await profileService.fetchUser();

      currentUser = {
        id: userData.data.id,
        ...userData.data.attributes,
      };
    } catch (error) {
      console.log(error);
    }
  };

  useEffect(() => {
    updateBonusesList();
    updateCurrentUser();
    dispatch({ type: 'UPDATE_POINTS', payload: currentUser });
  }, []);

  const onSuccess = () => {
    updateBonusesList();
    updateCurrentUser();
    dispatch({ type: 'UPDATE_POINTS', payload: currentUser });
  };

  return (
    <BonusContentWrapper>
      <MyBonuses>{state.currentUser['allowance-balance']} points to give away</MyBonuses>
      <SendBonusForm onSuccess={onSuccess} authService={authService} bonusService={bonusService} />
      <BonusList bonusList={bonusList} hasError={hasError} isLoading={isLoading} />
    </BonusContentWrapper>
  );
};

export default BonusContent;
