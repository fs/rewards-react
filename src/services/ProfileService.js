import api from './ApiService';
import AuthService from './AuthService';

const profileUrl = '/user/profile';

export default class ProfileService {
  static async fetchUser() {
    const token = AuthService.getToken();
    const config = {
      headers: { Authorization: `Bearer ${token}` },
    };
    const response = await api.get(profileUrl, config);
    return response.data;
  }
}
