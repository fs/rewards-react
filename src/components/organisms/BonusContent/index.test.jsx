import 'jest-dom/extend-expect';
import 'react-testing-library/cleanup-after-each';

import React from 'react';

import { render, waitForElement } from 'react-testing-library';
import 'jest-styled-components';

import BonusService from '../../../services/BonusService';
import AuthService from '../../../services/AuthService';
import ProfileService from '../../../services/ProfileService';
import BonusContent from '.';
import mockBonusResponse from '../../../mock_data/bonusResponse';

jest.mock('../../../services/BonusService');
jest.mock('../../../services/AuthService');
jest.mock('../../../services/ProfileService');

describe('BonusContent index', () => {
  let expectedToken;

  beforeEach(() => {
    expectedToken =
      'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1NTA3MzcwNjIsInN1YiI6MzczLCJ0eXBlIjoiYWNjZXNzIn0.JyTOZ8boBYlq0U3Iz3oVs7Tf-eeBLmD_Kl9ml2TO4YA';
    const mockGetToken = jest.fn(() => expectedToken);
    AuthService.getToken.mockImplementation(mockGetToken);
    jest.clearAllMocks();
    jest.resetModules();
  });

  test('render BonusContent correctly', () => {
    // Arrange
    // Act
    const { container } = render(<BonusContent />);
    // Assert
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render BonusList', async () => {
    // Arrange
    const expectedResponse = mockBonusResponse;
    const expectedUserData = {
      id: '39',
      type: 'users',
      attributes: {
        email: 'nadezhda.kharchuk@flatstack.com',
        'full-name': 'Nadezhda Kharchuk',
        username: 'nadezhda.kharchuk',
        'profile-image-avatar-url':
          'https://d14n4flg6h47l7.cloudfront.net/store/user/39/profile_image/avatar-58e3224793345ae993f75a6f373f3567.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIJWLCQQNSWJW7QQA%2F20190528%2Feu-central-1%2Fs3%2Faws4_request&X-Amz-Date=20190528T082822Z&X-Amz-Expires=900&X-Amz-SignedHeaders=host&X-Amz-Signature=9d3bd2740547bbffd76501b63526d4ae402e1afcc7be5584d1a2de0aaa5e7dab',
        'bonus-balance': 79,
        'allowance-balance': 469,
      },
    };

    const mockFetchBonusesList = jest.fn(
      () =>
        new Promise(resolve => {
          resolve({ data: expectedResponse });
        }),
    );

    BonusService.fetchBonusesList.mockImplementation(mockFetchBonusesList);

    const mockFetchUser = jest.fn(
      () =>
        new Promise(resolve => {
          resolve({ data: expectedUserData });
        }),
    );

    ProfileService.fetchUser.mockImplementation(mockFetchUser);

    // Act
    const { getByTestId } = render(<BonusContent />);

    const bonusList = await waitForElement(() => getByTestId('test-bonus-list'));

    // Assert
    expect(AuthService.getToken).toHaveBeenCalled();
    expect(ProfileService.fetchUser).toHaveBeenCalled();
    expect(BonusService.fetchBonusesList).toHaveBeenCalledWith(expectedToken);
    expect(bonusList).toBeInTheDocument();
  });

  test('should render error message if error', async () => {
    // Arrange
    const expectedError = new Error();

    const mockFetchBonusesList = jest.fn(
      () =>
        new Promise((resolve, reject) => {
          reject(expectedError);
        }),
    );

    BonusService.fetchBonusesList.mockImplementation(mockFetchBonusesList);

    // Act
    const { getByTestId } = render(<BonusContent />);
    const error = await waitForElement(() => getByTestId('test-error'));

    // Assert
    expect(error).toBeInTheDocument();
    expect(BonusService.fetchBonusesList).toBeCalledWith(expectedToken);
  });
});
