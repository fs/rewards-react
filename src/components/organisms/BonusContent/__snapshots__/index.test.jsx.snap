// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BonusContent index render BonusContent correctly 1`] = `
.c4 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  overflow: hidden;
  background-color: #d8d8d8;
  margin-right: 10px;
  cursor: pointer;
}

.c4:hover,
.c4.active {
  background-color: rgba(0,85,112,0.7);
}

.c4:last-child {
  margin-right: 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 20px;
}

.c5 {
  background-color: #63bc36;
  color: #fff;
  border-radius: 100px;
  height: 40px;
  line-height: 40px;
  padding: 0 20px;
  font-size: 16px;
  font-weight: bold;
  border: none;
}

.c5:hover,
.c5:focus {
  outline: none;
  background-color: #4f962b;
}

.c2 {
  width: 100%;
  background-color: #fff;
  border-radius: 5px;
  margin: 0 0 20px;
  padding: 20px;
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.1);
}

.c6 {
  margin: 20px;
  text-align: center;
}

.c0 {
  position: relative;
  margin-bottom: 1.25rem;
  margin-right: 2rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border-radius: 4px;
}

.c1 {
  position: absolute;
  top: -55px;
  line-height: 1.625rem;
  font-weight: 800;
  font-size: 1.375rem;
  color: #fff;
}

<div
  class="c0"
>
  <h2
    class="c1"
  >
     
    points to give away
  </h2>
  <form
    class="c2"
    data-testid="test-bonus-form"
  >
    <div
      data-testid="test-textarea-wrapper"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <img
            alt="Points"
            src="[object Object]"
          />
        </div>
        <div
          class="c4"
        >
          <img
            alt="User"
            src="[object Object]"
          />
        </div>
        <div
          class="c4"
        >
          <img
            alt="Hashtag"
            src="[object Object]"
          />
        </div>
      </div>
      <div
        class="rta  "
      >
        <textarea
          class="rta__textarea autocomplete-textarea"
          data-testid="test-textarea"
          name="bonus-text"
          placeholder="+100 @person add description for #create_awesomness"
        />
      </div>
    </div>
    <div
      data-testid="test-error-container"
    />
    <button
      class="c5"
      data-testid="test-button"
    >
      Give
    </button>
  </form>
  <div>
    <div
      class="c6"
      data-testid="test-loader"
    >
      Content is loading
    </div>
  </div>
</div>
`;
